Modules/mmapmodule.c

```
@@ -1117,6 +1117,46 @@ mmap_mmap_seek_impl(mmap_object *self, Py_ssize_t dist, int how)
     return NULL;
 }
 
+/*clinic*/
+
+/*[clinic input]
+mmap.mmap.set_name
+
+    name: str
+    /
+
+[clinic start generated code]*/
+
+static PyObject *
+mmap_mmap_set_name_impl(mmap_object *self, const char *name)
+/*[clinic end generated code: output=1edaf4fd51277760 input=6c7dd91cad205f07]*/
+{
+#if defined(MAP_ANONYMOUS) && defined(__linux__)
+    const char *prefix = "cpython:mmap:";
+    if (strlen(name) + strlen(prefix) > 79) {
+        PyErr_SetString(PyExc_ValueError, "name is too long");
+        return NULL;
+    }
+    if (self->flags & MAP_ANONYMOUS) {
+        char buf[80];
+        sprintf(buf, "%s%s", prefix, name);
+        _PyAnnotateMemoryMap(self->data, self->size, buf);
```

I know that this function suppresses `errno` but for the (Python) user, wouldn't it be better to know when the name couldn't be set? (maybe `_PyAnnotateMemoryMap` should return 0/1 depending on whether the errno was set?)

---

@picnixz Do you want emit warn log?

---

Oh right, I expect an `OSError` if `prctl()` fails. You can add a parameter to _PyAnnotateMemoryMap() to decide if an exception should be raised, or if the error should be ignored silently.

---

Yes, and you should document that OSError may be raised in this case (and not ValueError). Likewise, if the length is too long it's a ValueError that is raised.

---

I am not sure about raising an OSError, because if prctl() fails, there is no real side effect. Nothing actually happens.
What about the unit tests? Should we ignore OSError there as well? If someone wants to use this feature, do they need to catch OSError every time just because the kernel does not support it?
Because `_PyAnnotateMemoryMap ` does not raise any error where the platform is not compatible.
https://github.com/python/cpython/blob/16a305f15242ed2acac9bfdbb834804a00f6d9c5/Include/internal/pycore_mmap.h#L36-L40
I am more inclined to silently ignore the failure and emit a message to stderr instead.

---

I don't think `stderr` is a good choice -- it's hard to get info back from it, and it's hard to disable for users that don't want the spam.
Maybe the function could return the name that it set (possibly truncated, and with the common prefix), or `None`? That would be some actionable information that interested users could check.

Or make it like thread names, see the issue :)

---

Ignoring silently errors is not a good practice, it doesn't help to understand why the function failed. For tests, we can wait for the first CI failure to decide how to handle errors.

`_thread._set_name()` does raise OSError exceptions.

---

> Maybe the function could return the name that it set (possibly truncated, and with the common prefix), or None? That would be some actionable information that interested users could check.

This sounds great.

> For tests, we can wait for the first CI failure to decide how to handle errors.

All Fedora tests will be failed.

---

> All Fedora tests will be failed.

I wrote https://github.com/corona10/cpython/pull/65 to report the error to the caller. I updated the test to pass on Fedora.

---

Thank you Victor, I merged the PR!